### ISSUE TITLE
icclim cannot deal with precipitation unit (and variable name) from E-OBS dataset

### ISSUE THREAD

**Issue by pagecp** [cerfacs-globc/icclim#195]
* icclim version: 5.2.1
* Python version: 3.8

### Description

To calculate indices using precipitation, icclim expects precipitation variable pr, with name precipitation_flux and unit kg m-2 s-2, but E-OBS has precipitation variable rr, with name thickness_of_rainfall_amount and units mm. icclim aborts because of the units.

### Minimal reproducible example

`heavy_precipitation_days = icclim.r10mm(in_files=listOfFileNames, slice_mode='year')`

### Output received

```
2022-07-25 11:06:45,992 --- icclim 5.2.1
2022-07-25 11:06:45,993 --- BEGIN EXECUTION
2022-07-25 11:06:45,994 Processing: 0%
2022-07-25 11:06:46,100 Calculating climate index: R10mm
/Users/page/miniconda3/envs/icclimv5/lib/python3.8/site-packages/xclim/core/cfchecks.py:43: UserWarning: Variable has a non-conforming standard_name: Got `thickness_of_rainfall_amount`, expected `['precipitation_flux']`
  check_valid(vardata, "standard_name", data["standard_name"])
/Users/page/miniconda3/envs/icclimv5/lib/python3.8/site-packages/xclim/core/units.py:688: UserWarning: Data units millimeter are not compatible with requested [precipitation].
  check_units(val, bound_units.arguments.get(name, None))
---------------------------------------------------------------------------
DimensionalityError                       Traceback (most recent call last)
Input In [10], in <cell line: 5>()
      1 # Add specific unit spelling
      2 #units.define("@alias degC = C = deg_C = Celsius")
      3 
      4 # Add out_file='out_icclim.nc' to also output data into a NetCDF file
----> 5 heavy_precipitation_days = icclim.r10mm(in_files=listOfFileNames, slice_mode='year')

File ~/miniconda3/envs/icclimv5/lib/python3.8/site-packages/icclim/_generated_api.py:2547, in r10mm(in_files, var_name, slice_mode, time_range, out_file, ignore_Feb29th, netcdf_version, logs_verbosity)
   2494 def r10mm(
   2495     in_files: str | list[str] | Dataset | DataArray,
   2496     var_name: str | list[str] | None = None,
   (...)
   2502     logs_verbosity: Verbosity | str = Verbosity.LOW,
   2503 ) -> Dataset:
   2504     """
   2505     R10mm: Number of heavy precipitation days (Precip >=10mm)
   2506 
   (...)
   2545 
   2546     """
-> 2547     return icclim.index(
   2548         index_name="R10MM",
   2549         in_files=in_files,
   2550         var_name=var_name,
   2551         slice_mode=slice_mode,
   2552         time_range=time_range,
   2553         out_file=out_file,
   2554         ignore_Feb29th=ignore_Feb29th,
   2555         netcdf_version=netcdf_version,
   2556         logs_verbosity=logs_verbosity,
   2557     )

File ~/miniconda3/envs/icclimv5/lib/python3.8/site-packages/icclim/main.py:262, in index(in_files, index_name, var_name, slice_mode, time_range, out_file, threshold, callback, callback_percentage_start_value, callback_percentage_total, base_period_time_range, window_width, only_leap_years, ignore_Feb29th, interpolation, out_unit, netcdf_version, user_index, save_percentile, logs_verbosity, indice_name, user_indice, transfer_limit_Mbytes)
    260     result_ds = _compute_user_index_dataset(config=config, user_index=user_index)
    261 else:
--> 262     result_ds = _compute_ecad_index_dataset(
    263         config=config,
    264         index=index,
    265         threshold=threshold,
    266         current_history=ds.attrs.get("history", None),
    267     )
    268 if reset_coords_dict:
    269     result_ds = result_ds.rename(reset_coords_dict)

File ~/miniconda3/envs/icclimv5/lib/python3.8/site-packages/icclim/main.py:344, in _compute_ecad_index_dataset(config, index, threshold, current_history)
    342 else:
    343     config.threshold = threshold
--> 344     result_ds = _compute_ecad_index(index, config, current_history)
    345 return result_ds

File ~/miniconda3/envs/icclimv5/lib/python3.8/site-packages/icclim/main.py:397, in _compute_ecad_index(index, config, former_history)
    395 logging.info(f"Calculating climate index: {index.short_name}")
    396 result_ds = Dataset()
--> 397 res = index.compute(config)
    398 if isinstance(res, tuple):
    399     da, per = res

File ~/miniconda3/envs/icclimv5/lib/python3.8/site-packages/icclim/models/ecad_indices.py:312, in EcadIndex.<lambda>(c)
    298 SDII = (
    299     "SDII",
    300     lambda c: sdii(c),
    301     IndexGroup.RAIN,
    302     [PR],
    303 )
    304 CWD = (
    305     "CWD",
    306     lambda c: cwd(c),
    307     IndexGroup.RAIN,
    308     [PR],
    309 )
    310 R10MM = (
    311     "R10mm",
--> 312     lambda c: r10mm(c),
    313     IndexGroup.RAIN,
    314     [PR],
    315 )
    316 R20MM = (
    317     "R20mm",
    318     lambda c: r20mm(c),
    319     IndexGroup.RAIN,
    320     [PR],
    321 )
    322 RX1DAY = (
    323     "RX1day",
    324     lambda c: rx1day(c),
    325     IndexGroup.RAIN,
    326     [PR],
    327 )

File ~/miniconda3/envs/icclimv5/lib/python3.8/site-packages/icclim/ecad_functions.py:271, in r10mm(config)
    270 def r10mm(config: IndexConfig) -> DataArray:
--> 271     result = atmos.wetdays(
    272         config.pr.study_da, thresh="10 mm/day", freq=config.freq.panda_freq
    273     )
    274     return result

File ~/miniconda3/envs/icclimv5/lib/python3.8/site-packages/xclim/core/indicator.py:790, in Indicator.__call__(self, *args, **kwds)
    788     elif nm not in compute_das and nm in params:
    789         kwargs[nm] = params[nm]
--> 790 outs = self.compute(**compute_das, **kwargs, **var_kwargs)
    792 if isinstance(outs, DataArray):
    793     outs = [outs]

File <boltons.funcutils.FunctionBuilder-40>:2, in wetdays(pr, thresh, freq)

File ~/miniconda3/envs/icclimv5/lib/python3.8/site-packages/xclim/core/units.py:690, in declare_units.<locals>.dec.<locals>.wrapper(*args, **kwargs)
    687 for name, val in bound_args.arguments.items():
    688     check_units(val, bound_units.arguments.get(name, None))
--> 690 out = func(*args, **kwargs)
    692 # Perform very basic sanity check on the output.
    693 # Indice are responsible for unit management.
    694 # If this fails, it's a developer's error.
    695 if isinstance(out, tuple):

File ~/miniconda3/envs/icclimv5/lib/python3.8/site-packages/xclim/indices/_threshold.py:1771, in wetdays(pr, thresh, freq)
   1740 @declare_units(pr="[precipitation]", thresh="[precipitation]")
   1741 def wetdays(
   1742     pr: xarray.DataArray, thresh: str = "1.0 mm/day", freq: str = "YS"
   1743 ) -> xarray.DataArray:
   1744     """Wet days.
   1745 
   1746     Return the total number of days during period with precipitation over threshold.
   (...)
   1769     >>> wd = wetdays(pr, thresh="5 mm/day", freq="QS-DEC")
   1770     """
-> 1771     thresh = convert_units_to(thresh, pr, "hydro")
   1773     wd = threshold_count(pr, ">=", thresh, freq)
   1774     return to_agg_units(wd, pr, "count")

File ~/miniconda3/envs/icclimv5/lib/python3.8/site-packages/xclim/core/units.py:296, in convert_units_to(source, target, context)
    294     q = str2pint(source)
    295     # Return magnitude of converted quantity. This is going to fail if units are not compatible.
--> 296     return q.to(tu).m
    298 if isinstance(source, units.Quantity):
    299     return source.to(tu).m

File ~/miniconda3/envs/icclimv5/lib/python3.8/site-packages/pint/quantity.py:741, in Quantity.to(self, other, *contexts, **ctx_kwargs)
    724 """Return Quantity rescaled to different units.
    725 
    726 Parameters
   (...)
    737 pint.Quantity
    738 """
    739 other = to_units_container(other, self._REGISTRY)
--> 741 magnitude = self._convert_magnitude_not_inplace(other, *contexts, **ctx_kwargs)
    743 return self.__class__(magnitude, other)

File ~/miniconda3/envs/icclimv5/lib/python3.8/site-packages/pint/quantity.py:690, in Quantity._convert_magnitude_not_inplace(self, other, *contexts, **ctx_kwargs)
    687     with self._REGISTRY.context(*contexts, **ctx_kwargs):
    688         return self._REGISTRY.convert(self._magnitude, self._units, other)
--> 690 return self._REGISTRY.convert(self._magnitude, self._units, other)

File ~/miniconda3/envs/icclimv5/lib/python3.8/site-packages/pint/registry.py:1036, in BaseRegistry.convert(self, value, src, dst, inplace)
   1033 if src == dst:
   1034     return value
-> 1036 return self._convert(value, src, dst, inplace)

File ~/miniconda3/envs/icclimv5/lib/python3.8/site-packages/pint/registry.py:1951, in ContextRegistry._convert(self, value, src, dst, inplace)
   1947             src = self._active_ctx.transform(a, b, self, src)
   1949         value, src = src._magnitude, src._units
-> 1951 return super()._convert(value, src, dst, inplace)

File ~/miniconda3/envs/icclimv5/lib/python3.8/site-packages/pint/registry.py:1558, in NonMultiplicativeRegistry._convert(self, value, src, dst, inplace)
   1553     raise DimensionalityError(
   1554         src, dst, extra_msg=f" - In destination units, {ex}"
   1555     )
   1557 if not (src_offset_unit or dst_offset_unit):
-> 1558     return super()._convert(value, src, dst, inplace)
   1560 src_dim = self._get_dimensionality(src)
   1561 dst_dim = self._get_dimensionality(dst)

File ~/miniconda3/envs/icclimv5/lib/python3.8/site-packages/pint/registry.py:1069, in BaseRegistry._convert(self, value, src, dst, inplace, check_dimensionality)
   1066     # If the source and destination dimensionality are different,
   1067     # then the conversion cannot be performed.
   1068     if src_dim != dst_dim:
-> 1069         raise DimensionalityError(src, dst, src_dim, dst_dim)
   1071 # Here src and dst have only multiplicative units left. Thus we can
   1072 # convert with a factor.
   1073 factor, _ = self._get_root_units(src / dst)

DimensionalityError: Cannot convert from 'millimeter / day' ([length] / [time]) to 'millimeter' ([length])
```

**Comment by bzah** [comment]
In order to fix this we must detect that the input variable is a valid precipitation **accumulation** over time.
We can either use:
- `cell-method`, if it is a `time: sum` it's an accumulation.
- variable name (or `standard_name` attribute), if it is `thickness_of_rainfall_amount`, thickness can be interpreted as a accumulation (as opposed to a `precipitation_flux`).

Then we can infer the input time frequency `source_freq`.

Finally, before giving the DataArray to xclim we would turn the "units" into `mm / {source_freq}`.

**Comment by bzah** [comment]
xref: https://github.com/Ouranosinc/xclim/issues/1205

**Comment by bzah** [comment]
Pr in progress on xclim: 
I will duplicate the code into icc just so we can get this done for icc v6.0.
Once xclim 0.39 will be release, we will need to remove the duplication and upgrade our dependencies.

**Comment by bzah** [comment]
Fixed by #232
